### PR TITLE
Improvements to token counting for images and type annotations

### PR DESCRIPTION
**Pull request by pamelafox**
## Purpose

The original goal of this pull request was just to improve our type annotations to use `Mapping` instead of `dict` where we do not need mutability.

In writing the tests for that, I discovered that our token counting logic for images was never completed, so I finished its implementation and added tests.

## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test

* Run the new tests
* Use GPT-4-vision in developer settings, try sample questions.

### CODE REVIEW

File: app/backend/core/modelhelper.py [hunk 1]
```diff
@@ -39,14 +43,19 @@ def num_tokens_from_messages(message: dict[str, str], model: str) -> int:
...
     num_tokens = 2  # For "role" and "content" keys
-    for key, value in message.items():
+    for value in message.values():
         if isinstance(value, list):
-            for v in value:
-                # TODO: Update token count for images https://github.com/openai/openai-cookbook/pull/881/files
-                if isinstance(v, str):
-                    num_tokens += len(encoding.encode(v))
-        else:
+            for item in value:
```
Comment (pamelafox) [inline app/backend/core/modelhelper.py:48]: From https://github.com/openai/openai-cookbook/pull/881/files

File: app/backend/core/imageshelper.py [hunk 2]
```diff
@@ -34,3 +38,49 @@ async def fetch_image(blob_container_client: ContainerClient, result: Document)
         else:
             return None
     return None
+
+
+def get_image_dims(image):
```
Comment (pamelafox) [inline app/backend/core/imageshelper.py:43]: These functions are taken from https://github.com/openai/openai-cookbook/pull/881/files

File: app/backend/core/imageshelper.py [hunk 3]
```diff
@@ -1,8 +1,12 @@
 import base64
+import math
 import os
+import re
+from io import BytesIO
 from typing import Optional
 
 from azure.storage.blob.aio import ContainerClient
+from PIL import Image
```
Comment (pamelafox) [inline app/backend/core/imageshelper.py:9]: This does add a pillow dependency :(

File: app/backend/approaches/chatapproach.py [hunk 4]
```diff
@@ -110,7 +110,7 @@ def get_messages_from_history(
 
         total_token_count = 0
         for existing_message in message_builder.messages:
-            total_token_count += message_builder.count_tokens_for_message(dict(existing_message))  # type: ignore
+            total_token_count += message_builder.count_tokens_for_message(existing_message)
```
Comment (pamelafox) [inline app/backend/approaches/chatapproach.py:113]: Lol, the original goal of this PR was just to be able to remove `type: ignore`. Success!

File: app/backend/core/imageshelper.py (hunk 3 above)
Comment (pamelafox) [reply app/backend/core/imageshelper.py:9]: We do already have pillow in scripts, so not a big deal, but I'd feel better about it if it was required for non-vision use cases. I guess we already have precedent for that though, with all our authentication related dependencies.

File: app/backend/core/imageshelper.py (hunk 2 above)
Comment (mattgotteiner) [reply app/backend/core/imageshelper.py:43]: do you mind adding this a s a comment?

File: app/backend/core/modelhelper.py (hunk 1 above)
Comment (mattgotteiner) [reply app/backend/core/modelhelper.py:48]: same - please cite this in a comment

**Review by mattgotteiner — APPROVED** [review]

**Pull request merged**